rabbitmq/src/main/java/io/micronaut/rabbitmq/connect/RabbitConnectionFactory.java

```diff
@@ -49,27 +51,53 @@ public class RabbitConnectionFactory {
      * @param connectionFactory The factory to create the connection
      * @param beanContext The bean context to dynamically retrieve the executor service
      * @return The connection
+     * @deprecated Use {@link RabbitConnectionFactory#connection(RabbitConnectionFactoryConfig, TemporarilyDownConnectionManager, BeanContext)}
+     */
+    @Deprecated(since = "4.2.0", forRemoval = true)
+    Connection connection(RabbitConnectionFactoryConfig connectionFactory,
+                          BeanContext beanContext) {
+        return connection(connectionFactory, beanContext.getBean(TemporarilyDownConnectionManager.class), beanContext);
+    }
+
+    /**
+     * @param connectionFactory The factory to create the connection
+     * @param temporarilyDownConnectionManager The temporarily down connection manager
+     * @param beanContext The bean context to dynamically retrieve the executor service
+     * @return The connection
+     * @since 4.2.0
      */
     @Singleton
     @EachBean(RabbitConnectionFactoryConfig.class)
     Connection connection(RabbitConnectionFactoryConfig connectionFactory,
+                          TemporarilyDownConnectionManager temporarilyDownConnectionManager,
                           BeanContext beanContext) {
         try {
-            ExecutorService executorService = beanContext.getBean(ExecutorService.class, Qualifiers.byName(connectionFactory.getConsumerExecutor()));
-            Optional<List<Address>> addresses = connectionFactory.getAddresses();
-            Connection connection;
-            if (addresses.isPresent()) {
-                connection = connectionFactory.newConnection(executorService, addresses.get());
-            } else {
-                connection = connectionFactory.newConnection(executorService);
-            }
+            Connection connection = newConnection(connectionFactory, temporarilyDownConnectionManager, beanContext);
             activeConnections.add(new ActiveConnection(connection, connectionFactory));
             return connection;
         } catch (IOException | TimeoutException e) {
             throw new BeanInstantiationException("Error creating connection to RabbitMQ", e);
         }
     }
 
+    private Connection newConnection(RabbitConnectionFactoryConfig factory, TemporarilyDownConnectionManager temporarilyDownConnectionManager, BeanContext context) throws IOException, TimeoutException {
+        ExecutorService executorService = context.getBean(ExecutorService.class, Qualifiers.byName(factory.getConsumerExecutor()));
+        Optional<List<Address>> addresses = factory.getAddresses();
+        try {
+            if (addresses.isPresent()) {
+                return factory.newConnection(executorService, addresses.get());
+            }
+            return factory.newConnection(executorService);
+        } catch (ConnectException e) {
+            // Check if automatic recovery is enabled
+            if (factory.isAutomaticRecoveryEnabled()) {
```

Comment:
Are you sure that's how it works? Looking at the code if `automaticRecovery` is enabled it should already return an instance of `AutorecoveringConnection`

Reply:
> Are you sure that's how it works? Looking at the code if `automaticRecovery` is enabled it should already return an instance of `AutorecoveringConnection`

@dstepanov `com.rabbitmq.client.ConnectionFactory` only returns an `AutorecoveringConnection` if `automaticRecovery` is enabled and the RabbitMQ server is up when the connection is created. If you look in depth at [the source code](https://github.com/rabbitmq/rabbitmq-java-client/blob/main/src/main/java/com/rabbitmq/client/ConnectionFactory.java#L1268), you'll see this:

```java
if (isAutomaticRecoveryEnabled()) {
    AutorecoveringConnection conn = new AutorecoveringConnection(
        params, fhFactory, addressResolver, metricsCollector, observationCollector);
    conn.init(); // <---- THIS LINE THROWS AN EXCEPTION IF THE SERVER IS DOWN
    return conn;
} else ...
```

In practice, RabbitMQ auto recovering connections only work if they can connect to the server when you create them. That's why I had to create the "temporarily down" connection which extends from `AutorecoveringConnection`.